File: Sep/Day15/maximum-number-of-words-you-can-type.js
//  Time: O(N * B)  ::  Space:  O(1)
class Solution {
    canBeTypedWords(text, brokenLetters) {
        let count = 0;
        let words = text.split(" ");
        for (let word of words) {
            count++;
            for (let x of brokenLetters) {
                if (word.includes(x)) {
                    count--;
                    break;
                }
            }
        }
        return count;
    }
}




//  Time: O(N + B)  ::  Space:  O(1)
class Solution {
    canBeTypedWords(text, brokenLetters) {
        let broken = new Set(brokenLetters);
        let count = 0;
        for (let word of text.split(" ")) {
            let isBroken = false;
            for (let ch of word) {
                if (broken.has(ch)) {
                    isBroken = true;
                    break;
                }
            }
            if (!isBroken) count++;
        }
        return count;
    }
}




//  Time: O(N)  ::  Space:  O(1)
class Solution {
    canBeTypedWords(text, brokenLetters) {
        let brokenArr = Array(26).fill(0);
        for (let ch of brokenLetters) {
            brokenArr[ch.charCodeAt(0) - 97] = 1;
        }

        let wordIsBroken = false;
        let count = 1;
        for (let ch of text) {
            if (ch === ' ') {
                count++;
                wordIsBroken = false;
            } else {
                if (!wordIsBroken && brokenArr[ch.charCodeAt(0) - 97] === 1) {
                    wordIsBroken = true;
                    count--;
                }
            }
        }
        return count;
    }
}

File: Sep/Day15/maximum-number-of-words-you-can-type.java
//  Time: O(N * B)  ::  Space:  O(1)
class Solution {
    public int canBeTypedWords(String text, String brokenLetters) {
        int count = 0;
        String[] words = text.split(" ");
        for (String word : words) {
            count++;
            for (char x : brokenLetters.toCharArray()) {
                if (word.indexOf(x) != -1) {
                    count--;
                    break;
                }
            }
        }
        return count;
    }
}



//  Time: O(N + B)  ::  Space:  O(1)
class Solution {
    public int canBeTypedWords(String text, String brokenLetters) {
        Set<Character> broken = new HashSet<>();
        for (char ch : brokenLetters.toCharArray()) {
            broken.add(ch);
        }
        int count = 0;
        for (String word : text.split(" ")) {
            boolean isBroken = false;
            for (char ch : word.toCharArray()) {
                if (broken.contains(ch)) {
                    isBroken = true;
                    break;
                }
            }
            if (!isBroken) count++;
        }
        return count;
    }
}



//  Time: O(N)  ::  Space:  O(1)
class Solution {
    public int canBeTypedWords(String text, String brokenLetters) {
        int[] brokenArr = new int[26];
        for (char ch : brokenLetters.toCharArray()) {
            brokenArr[ch - 'a'] = 1;
        }
        
        boolean wordIsBroken = false;
        int count = 1;
        for (char ch : text.toCharArray()) {
            if (ch == ' ') {
                count++;
                wordIsBroken = false;
            } else {
                if (!wordIsBroken && brokenArr[ch - 'a'] == 1) {
                    wordIsBroken = true;
                    count--;
                }
            }
        }
        return count;
    }
}
